File: apps/opentelemetry/test/attributes_sampler.erl
%% sampler returns `DROP' if a span attribute matches the configured attribute map
%% otherwise `RECORD_AND_SAMPLE'
-module(attributes_sampler).

-behavior(otel_sampler).

-export([description/1,
         setup/1,
         should_sample/7]).

-include("otel_sampler.hrl").

setup(Attributes) when is_map(Attributes) ->
    Attributes;
setup(_) ->
    #{}.

description(_) ->
    <<"AttributesSampler">>.

-spec should_sample(Ctx, TraceId, Links, SpanName, SpanKind, Attributes, Config) -> Result
              when Ctx :: otel_ctx:t(),
                   TraceId :: opentelemetry:trace_id(),
                   Links :: otel_links:t(),
                   SpanName :: opentelemetry:span_name(),
                   SpanKind :: nopentelemetry:span_kind(),
                   Attributes :: opentelemetry:attributes_map(),
                   Config :: otel_sampler:sampler_config(),
                   Result :: otel_sampler:sampling_result().
should_sample(_Ctx, _TraceId, _Links, _SpanName, _SpanKind, Attributes, ConfigAttributes) ->
    case has_match(Attributes, ConfigAttributes) of
        true ->
            {?DROP, [], []};
        _ ->
            {?RECORD_AND_SAMPLE, [], []}
    end.

has_match(A, B) ->
    {Min, Max} = if
                     map_size(A) < map_size(B) ->
                         {A, B};
                     true ->
                         {B, A}
                 end,
    I = maps:iterator(Min),
    has_match_(maps:next(I), Max).

has_match_(none, _) ->
    false;
has_match_({K, V, _I}, B) when map_get(K, B) =:= V ->
    true;
has_match_({_K, _V, I}, B) ->
    has_match_(maps:next(I), B).

File: test/support/attributes_sampler.ex
defmodule AttributesSampler do
  def setup(attributes) when is_map(attributes) do
    attributes
  end

  def setup(_) do
    %{}
  end

  def description(_) do
    "AttributesSampler"
  end

  def should_sample(
        _ctx,
        _trace_id,
        _links,
        _span_name,
        _span_kind,
        attributes,
        config_attributes
      ) do
    case has_match(attributes, config_attributes) do
      true ->
        {:drop, [], []}

      _ ->
        {:record_and_sample, [], []}
    end
  end

  def has_match(a, b) do
    {min, max} =
      if map_size(a) < map_size(b) do
        {a, b}
      else
        {b, a}
      end

    i = :maps.iterator(min)
    has_match_(:maps.next(i), max)
  end

  def has_match_(:none, _), do: false
  def has_match_({k, v, _i}, b) when :erlang.map_get(k, b) === v, do: true
  def has_match_({_k, _v, i}, b), do: has_match_(:maps.next(i), b)
end
